## Problem Definition

This task demonstrates data handling and analysis using **pandas** in two parts:

### Part 1: Custom Dataset
- Build a DataFrame from a dictionary  
- it have **5 columns**, **15 rows**, and a **indexs** generated using a **vectorized approach**  rather than a manual loop.  
This ensures concise, efficient, and Pythonic code

### Part 2: Titanic Dataset
dataset source from: https://www.kaggle.com/datasets/yasserh/titanic-dataset?select=Titanic-Dataset.csv

**Problem Type**

The Titanic dataset is a **supervised classification problem**:

- **Supervised**  because we have labeled data (`Survived` column) that serves as the target variable.  
- **Classification**  because the target is categorical (0 = did not survive, 1 = survived). The task is to classify passengers into one of these two categories based on their features.

1. **Exploration**  `.head()`, `.info()`, `.describe()`  
2. **Data Cleaning**  Fill `Age` with median, `Embarked` with mode, drop `Cabin`, remove duplicates  
3. **Analysis**  `groupby()` for survival by gender, class, age group, average age per class  
4. **Filtering**  Female survivors, child survivors, 1st class survivors  
5. **Insights**  Who survived more based on gender, effect of class, children prioritized, best survival combination 

**key features:**
* PassengerId – Unique identifier for each passenger
* Survived – Survival status (0 = No, 1 = Yes)
* Pclass – Passenger class (1 = First, 2 = Second, 3 = Third)
* Name – Passenger name
* Sex – Gender of the passenger
* Age – Age of the passenger
* SibSp – Number of siblings/spouses aboard
* Parch – Number of parents/children aboard
* Ticket – Ticket number
* Fare – Ticket fare paid
* Cabin – Cabin number
* Embarked – Port of embarkation (C, Q, S)
  
**target variable** Survived (that indicates whether a passenger survived or not)
### Constraints
- Use **pandas only**  
- Code must be **clean and readable**  
- Handle **missing values properly**


In [75]:
# import pandas module for data manipulation and analysis
import pandas as pd

# Part 1: Create Your Own Dataset
Build a DataFrame from a Dictionary

In [76]:
# create dictionary
car = {
    "Car_Make": [
        "Toyota", "BMW", "Mercedes", "Honda", "Ford",
        "Chevrolet", "Audi", "Hyundai", "Kia", "Nissan",
        "Volkswagen", "Mazda", "Jeep", "Toyota", "Tesla"
    ],
    "Color": [
        "Red", "Black", "Silver", "Blue", "White",
        "Yellow", "Gray", "Green", "Black", "White",
        "Blue", "Red", "Green", "Silver", "White"
    ],
    "Odometer(km)": [
        25000, 18000, 30000, 22000, 40000,
        35000, 27000, 15000, 20000, 32000,
        28000, 24000, 41000, 19000, 12000
    ],
    "Doors": [
        4, 4, 4, 4, 2,
        2, 4, 4, 4, 4,
        4, 4, 4, 4, 4
    ],
    "Price($)": [
        20000, 55000, 42000, 21000, 35000,
        37000, 40000, 19000, 25000, 24000,
        23000, 28000, 32000, 27000, 45000
    ]
}

# Convert dictionary to DataFrame
car_make = pd.DataFrame(car, index=range(1,16))
car_make


,Car_Make,Color,Odometer(km),Doors,Price($)
1,Toyota,Red,25000,4,20000
2,BMW,Black,18000,4,55000
3,Mercedes,Silver,30000,4,42000
4,Honda,Blue,22000,4,21000
5,Ford,White,40000,2,35000
6,Chevrolet,Yellow,35000,2,37000
7,Audi,Gray,27000,4,40000
8,Hyundai,Green,15000,4,19000
9,Kia,Black,20000,4,25000
10,Nissan,White,32000,4,24000


# Part 2: Real Dataset — Titanic

In [77]:
# Load dataset
train=pd.read_csv("Titanic-Dataset.csv")


## Step 1: Exploration
initial Exploration(inspection)
* .head()
* .info()
* .describe()

In [78]:
# first 5 elements
train.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


* The first five rows of the Titanic dataset, showing passenger details such as ID, survival status, class, name, gender, age, family aboard, ticket information, fare, cabin, and port of embarkation.
* This preview helps confirm that the dataset is loaded correctly and provides an initial look at the structure of the data.

In [79]:
# loc access cause i want to know how is the data at the middle
train.loc[100:115]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
100,101,0,3,"Petranec, Miss. Matilda",female,28.0,0,0,349245,7.8958,NaN,S
101,102,0,3,"Petroff, Mr. Pastcho (""Pentcho"")",male,NaN,0,0,349215,7.8958,NaN,S
102,103,0,1,"White, Mr. Richard Frasar",male,21.0,0,1,35281,77.2875,D26,S
103,104,0,3,"Johansson, Mr. Gustaf Joel",male,33.0,0,0,7540,8.6542,NaN,S
104,105,0,3,"Gustafsson, Mr. Anders Vilhelm",male,37.0,2,0,3101276,7.9250,NaN,S
105,106,0,3,"Mionoff, Mr. Stoytcho",male,28.0,0,0,349207,7.8958,NaN,S
106,107,1,3,"Salkjelsvik, Miss. Anna Kristine",female,21.0,0,0,343120,7.6500,NaN,S
107,108,1,3,"Moss, Mr. Albert Johan",male,NaN,0,0,312991,7.7750,NaN,S
108,109,0,3,"Rekic, Mr. Tido",male,38.0,0,0,349249,7.8958,NaN,S
109,110,1,3,"Moran, Miss. Bertha",female,NaN,1,0,371110,24.1500,NaN,Q


* The `.loc[]` output displays passenger records from index 100 to 115.
* This slice provides a closer look at a subset of the dataset, useful for checking consistency, spotting missing values, and observing patterns across a smaller group of passengers.

In [80]:
# displays the Last 5 data
train.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


In [81]:
# it's tell me column information
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


* we have 891 rows and 12 columns
* 5 object types and 7 numerical data type
* also have missing values
* Approximate size of the DataFrame in memory.

In [82]:
#add missing values in each columns
train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

`Age`  column 177 missing values

`Cabin`  column 687 missing values morethan half of the data is missing.

`Embarked` column 2 missingness which is minimal

* Age greatly varies to for it's robustness of outlier fill with median
* since Embarked is catagorical fill with mode
* droping `Cabin` column makes sense because it’s mostly missing and not very informative for basic models unless extracted.

In [83]:
#summary statistic of all numeric columns
train.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


Count is numbes of non-null values `Age` columns 714 counts that means 177 data is missing.
* Survival rate is low **38%**, showing the tragedy’s scale.
* Class distribution is skewed toward 3rd class, which may important for survival analysis.
* Age distribution shows most passengers were young adults, but there were infants and elderly too.
* Fare distribution is **highly skewed**.
* Family features (SibSp, Parch) show most passengers traveled alone or with one relative

## Step 2: Data Cleaning
Handle missing values:
* Age → median
* Embarked → mode
* Cabin → drop
* remove duplicates if any

In [84]:
# check if there any outlier
train.duplicated().sum()

np.int64(0)

* it returns boolean if there is duplicated it returns True and if no returns False
* `.sum()` return 0 which indicates there is no dupplicated records. SO, **Each passenger entry is unique**

In [85]:
# i defined catagorical and numerical features for simplicity to  see which feature is catagorical for reference

catagorical=train.select_dtypes(include="object").columns
numerical=train.select_dtypes(exclude="object").columns

In [86]:
numerical

Index(['PassengerId', 'Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare'], dtype='object')

`Age` is numerical data type and have 177 missing values fill with mean or median so first let's check the distribution is there any outlier or not because it's matter alot.

**Because of the contraints use pandas only i can't plot the distribution. let me use IQR method to detect outlier**

In [87]:
#outlier detection usig IQR 
# define lower and upper bound to check the value is below the lower bound or above the upper bound

q1=train["Age"].quantile(0.25)
q3=train["Age"].quantile(0.75)
IQR=q3-q1
lower_bound=q1 - 1.5*IQR
upper_bound=q3 + 1.5*IQR
outlier=train[(train["Age"] < lower_bound) | (train["Age"] > upper_bound)]
outlier["Age"]

33     66.0
54     65.0
96     71.0
116    70.5
280    65.0
456    65.0
493    71.0
630    80.0
672    70.0
745    70.0
851    74.0
Name: Age, dtype: float64

`Age` columns have outliers filling with mean leads to `False` implication to the Age because mean is sensetive to outliers i should use median. 

In [88]:
# Age fill with pandas i can't use scikit to impute

train["Age"] = train["Age"].fillna(train["Age"].median())
train["Age"].isnull().sum()

np.int64(0)

Thanks for Pandas i easily impute the `Age` column. No missing values

In [89]:
catagorical

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')

* `Cabin` and `Embarked` columns have object data type.
* **decision** impute `Cabin`  with mode and drop the `Cabin` columns

In [90]:
# Embarked catagorical column with 2 miss we can fill with mode

train["Embarked"] = train["Embarked"].fillna(train["Embarked"].mode)
train["Embarked"].isnull().sum()

np.int64(0)

**all missings are handled with most frequent**

In [91]:
# Drop Cabin column
train.drop("Cabin",axis=1,inplace=True)

In [92]:
# Check if Cabin is still in the DataFrame(using membership `in`)

'Cabin' in train.columns

False

In [93]:
# display the column names
train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Embarked'],
      dtype='object')

The **`Cabin`** column successfully removed from the dataset.

## Step 3: Data Analysis
Using `groupby()`:
* Survival rate by gender
* Survival rate by class
* Average age per class
* Survival rate by age group


In [94]:
# Survival rate by gender
train.groupby("Sex")["Survived"].mean()

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64

**observation:**
* Survival rate of Female is 74%
* survival rate of male 19%
  
**Insights:**
- **Women had a much higher survival rate** compared to men.
- This reflects the "women and children first" Hypothesis.

**Decision:** Gender should be considered as a strong predictor in survival models.

In [95]:
# Survival rate by class
train.groupby("Pclass")["Survived"].mean()

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64

**observation**
Survival chances decreased with lower class.
* 1st class 63% the highest survival rate
* 2nd class 47% moderate survival rate.
* 3rd class 24% the lowest survival rate.
  
**Insights** 
* This highlights the **strong influence of socio-economic status** on survival chances. Higher‑class passengers had better access to lifeboats and crew assistance. 
 
**Decision:** Passenger class strongly influenced survival.so, `Pclass` is an important feature for predictive modeling.

In [96]:
# Average Age per Class
train.groupby("Pclass")["Age"].mean()

Pclass
1    36.812130
2    29.765380
3    25.932627
Name: Age, dtype: float64

**Observation:** Age distribution varies across classes.

**Insight:** Higher classes may have slightly older passengers.

**Decision:** Age combined with class could refine survival predictions.

In [97]:
# Survival Rate by Age Group
# first le's create Age group using `.cut()`

train["AgeGroup"] = pd.cut(train["Age"], bins=[0, 12, 18, 35, 60, 100],
                           labels=["Child", "Teen", "Adult", "Middle-aged", "Senior"])
# 
train.groupby("AgeGroup",observed=True)["Survived"].mean()       # in the future default observed= False will deprecated


AgeGroup
Child          0.579710
Teen           0.428571
Adult          0.353271
Middle-aged    0.400000
Senior         0.227273
Name: Survived, dtype: float64

**Observation:**
- Children had the highest survival rate ~58% among age groups.  
- Seniors had the lowest survival rate.  
- Teens, adults, and middle-aged passengers had moderate survival rates.  

**Insight:**
- **Younger passengers were prioritized** during evacuation, consistent with the "women and children first" policy.  
- Older passengers faced lower survival chances, possibly due to mobility challenges or evacuation priorities.  

**Decision:**
- Age grouping is a valuable feature for predictive modeling.  
- Combining  with others features will likely improve survival prediction accuracy in Modeling.

## Step 4: Filtering
* Female passengers who survived
* Children who survived
* Passengers in 1st class with high survival probability


In [98]:
# Female Passengers Who Survived
female_survivors = train[(train["Sex"] == "female") & (train["Survived"] == 1)]
female_survivors.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,AgeGroup
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,Middle-aged
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,Adult
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,Adult
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,S,Adult
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,C,Teen


**Observation:** many female passengers survived.

**Insight:** Reflects evacuation priority for women.

**Decision:** Gender is a strong predictor for survival.

In [99]:
# Children Who Survived
child_survivors = train[(train["AgeGroup"] == "Child") & (train["Survived"] == 1)]
child_survivors.head()


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,AgeGroup
10,11,1,3,"Sandstrom, Miss. Marguerite Rut",female,4.00,1,1,PP 9549,16.7000,S,Child
43,44,1,2,"Laroche, Miss. Simonne Marie Anne Andree",female,3.00,1,2,SC/Paris 2123,41.5792,C,Child
58,59,1,2,"West, Miss. Constance Mirium",female,5.00,1,2,C.A. 34651,27.7500,S,Child
78,79,1,2,"Caldwell, Master. Alden Gates",male,0.83,0,2,248738,29.0000,S,Child
125,126,1,3,"Nicola-Yarred, Master. Elias",male,12.00,1,0,2651,11.2417,C,Child


**Observation:** Children had relatively high survival rates.

**Insight:** Supports the "women and children first" policy.

**Decision:** Age grouping is useful for feature engineering.

In [100]:
# Passengers in 1st Class with High Survival Probability
first_class_high_survival = train[(train["Pclass"] == 1) & (train["Survived"] == 1)]
first_class_high_survival.head()


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,AgeGroup
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,Middle-aged
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,Adult
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,S,Middle-aged
23,24,1,1,"Sloper, Mr. William Thompson",male,28.0,0,0,113788,35.5000,S,Adult
31,32,1,1,"Spencer, Mrs. William Augustus (Marie Eugenie)",female,28.0,1,0,PC 17569,146.5208,C,Adult


**Observation:** 1st‑class passengers had the highest survival rate.

**Insight:** Socio‑economic status influenced survival chances.

**Decision:** Passenger class should be included as a key feature in predictive models.

## Step 5: Insights
Answer clearly:
1. Who was more likely to survive? (gender)
* Female passengers were far more likely to survive compared to males.  
* Reason: Evacuation followed the "women and children first" policy, giving women priority access to lifeboats.
2. Did class affect survival?
* Yes, Survival rates decreased with lower class.
  * 1st Class: 63%  
  * 2nd Class: 47%  
  * 3rd Class: 24%  
* Reason: Higher‑class passengers were located closer to lifeboats and received quicker assistance.
3. Were children prioritized?
* Yes, Children had the highest survival rate among age groups ~ 58%.  
*Reason: They were prioritized during evacuation alongside women, consistent with historical accounts.
5. What combination had the highest survival rate?
* Female passengers in 1st class had the highest survival probability.  
* Reason: Gender and class together offered the greatest advantage — social status and evacuation priority combined to maximize survival chances.

## EDA summary

* The exploratory data analysis of the Titanic dataset shows that survival was strongly influenced by gender, class, and age. Women were far more likely to survive than men, with about `74%` of female passengers rescued compared to only `19%` of males.

* Passenger class also played a decisive role: first-class travelers had the highest survival rate, followed by second-class, while third-class passengers faced the lowest chances.

* Age further shaped outcomes, as children were prioritized during evacuation and had the highest survival rate among age groups, whereas seniors had the lowest.

*  When these factors were combined, **female passengers in first class** emerged as the group with the greatest survival probability.
* Overall, survival on the Titanic was determined by the intersection of **social status, age, and evacuation policies**, providing valuable insights for predictive modeling and reflecting the historical realities of the disaster.
